File: binary-search-medium/74.搜索二维矩阵.py
#
# @lc app=leetcode.cn id=74 lang=python3
#
# [74] 搜索二维矩阵
#
from typing import List


# @lc code=start
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row, col = len(matrix), len(matrix[0])
        left, right = 0, row * col - 1
        while left <= right:
            mid = left + (right - left) // 2
            if matrix[mid // col][mid % col] < target:
                left = mid + 1
            elif matrix[mid // col][mid % col] > target:
                right = mid - 1
            else:
                return True
        return False

# @lc code=end

File: binary-search-medium/74.搜索二维矩阵.cpp
/*
 * @lc app=leetcode.cn id=74 lang=cpp
 *
 * [74] 搜索二维矩阵
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target) {
        int row = matrix.size(), col = matrix.size() == 0 ? 0 : matrix[0].size();
        if (row == 0 || col == 0) return false;
        int left = 0, right = row * col - 1;
        while (left <= right) {
            int mid = left + ((right - left) >> 1);
            int x = mid / col, y = mid % col;
            if (matrix[x][y] < target)
                left = mid + 1;
            else if (matrix[x][y] > target)
                right = mid - 1;
            else
                return true;
        }
        return false;
    };
};
// @lc code=end
